Constraints visualization

ojura opened the pull request:
(Sorry, I had to delete my repo and recreate this pull request. Here it is again.)

Originally by @jihoonl. Some nits by @wohe from #337 have been addressed. Some haven't (e.g. https://github.com/googlecartographer/cartographer_ros/pull/337#discussion_r117281636 & https://github.com/googlecartographer/cartographer_ros/pull/337#discussion_r117281982, https://github.com/googlecartographer/cartographer_ros/pull/337#discussion_r117280199). Disclaimer: I have only updated @jihoonl's code so that it compiles with regards to API changes, without giving much thought what it actually does. :)

ojura commented:
This should probably be refactored so it uses a single LINE_LIST marker instead of creating a marker for each constraint, which is very slow. Perhaps, for intra constraints, the new coloring system could be used to color contraints differently for each submap.

wohe commented on cartographer_ros/cartographer_ros/map_builder_bridge.cc:
```diff
@@ -246,6 +247,97 @@ visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodesList() {
+
+    constraint_marker.points.push_back(submap_point);
+    constraint_marker.points.push_back(submap_pose_point);
+    residual_error_marker.points.push_back(submap_pose_point);
```
I think I understand the `residual_error_marker`, but I'm not sure I understand the purpose of the `constraint_marker`. Could you add a comment about what these markers are showing?

ojura replied:
Pinging @jihoonl @domeinzer.

domeinzer replied:
The idea behind the constraint marker was to visualize the two points which are connected to each other (especially because we (Magazino) wanted to have constraints on landmarks and have a debugging tool for those constraints.
The residual error marker was just to visualize the remaining error after optimization.

jihoonl replied:
constraint marker visualizes connectivity between trajectory node and submap. It helps to understand which nodes belong to(or connected to) which submap while robot moves.

![constraints](https://user-images.githubusercontent.com/1556736/27546101-11c905a4-5a92-11e7-9971-f5dd948ba96e.png)

In the picture, each number represents origin of a submap(green: local_pose, red:optimized_pose). Yellow line indicates which trajectory node is connected to which submap as inter submap constraints.

Mofef replied:
You have a point there, @wohe. But maybe this explanation helps: Originally I wanted to see the two end points of a `constraint`. So the `submap` an the `trajectory_node`, because I was adding additional `constraints` and I wanted to see if it works.
Later I wanted to see where the `constraints` are "pulling" a `trajectory_node`. This is what the `residual_error_marker`. 
Since technically it is not only pulling a `trajectory_node` torwards a point on a `submap`, but also a `submap` torwards that `trajectory_node`, it might still be useful to see the `submap` a constraint is attached to (or: whether two `trajectory_nodes` are constrained to the same `submap`.) This is the `constraint_marker`. Now in most use cases the `residual_error_marker` might be the more practical, which is why I would keep them as seperate topics, but with Landmarks in the pipeline this might soon get more relevant again.

wohe commented:
Sorry, I did not read your comments first. Yes, I agree.

ojura commented on cartographer_ros/cartographer_ros/map_builder_bridge.cc:
```diff
@@ -246,6 +247,97 @@ visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodesList() {
+    if (constraint.tag ==
+        cartographer::mapping::SparsePoseGraph::Constraint::INTER_SUBMAP) {
+      // yellow
+      color_constraint.a = 1.0;
```
Sure, that would be a good idea, I'll do it. Do you want to hardcode alpha = 1 in GetColor()?

wohe replied:
Yes. Until we see a use case that needs a different alpha, let's hardcode it.

wohe commented:
@ojura Friendly ping. You're not blocked by anyone, right?

ojura commented:
Hi, sorry, I had some other stuff going on (and spent the bigger part of the last week fixing CMake build errors). I definitely have finishing this on my todo list for this week. If anyone wants to collaborate, feel free to post here or to send me a pull request.

> You're not blocked by anyone, right?
 
I am not sure what you mean :)

wohe commented:
No problem and no need to hurry. I just wanted to make sure you are not waiting on anyone else to do something before you can continue.

Mofef commented on cartographer_ros/cartographer_ros/map_builder_bridge.cc:
```diff
@@ -246,6 +247,97 @@ visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodesList() {
+      color_error.b = 1.0;
+    }
+
+    constraint_marker = createVisualizationMarker(
```
I agree on the first part.
But I wouldn't merge the constraint markers and the residual markers into one marker, since then you couldn't disable them individually. Or am I mistaken about this?

ojura replied:
They could be put inside a single marker topic, but inside separate namespaces, similar to how you separated residual and constraint markers.

jihoonl commented:
@ojura , @Mofef wanted to create a PR #400 for your fork([larics:constraints_visualization](https://github.com/larics/cartographer_ros/tree/constraints_visualization)). But, somehow, he couldn't create a PR against your repo. Can you apply the changes in #400 in your branch? then we will close #400.

ojura commented:
Now it's working again same as before. Putting everything inside a single marker and coloring is not yet done.

ojura commented:
Hello everyone (@wohe @Mofef @jihoonl), please have a look. I completed the refactoring and added coloring. Everything is now published on a single topic, but in separate namespaces, so you can still disable everything.

For intra constraints, I added different coloring for different submaps, while the residuals are now bright red. For inter constraints, I retained Magazino's fixed bright yellow/teal color scheme.

ojura commented:
By the way, it seems the pixel bleaching code in shaders of the Submaps plugin sometimes clashes with rviz's rendering, which affects the LINE_LIST marker - especially in places where more that 2 submaps are overlaid.

https://www.youtube.com/watch?v=_XMGt5iYzdI

Notice how it affects the grid too.

wohe commented on cartographer_ros/cartographer_ros/node_constants.h:
```diff
@@ -33,6 +33,8 @@ constexpr char kSubmapQueryServiceName[] = "submap_query";
 constexpr char kStartTrajectoryServiceName[] = "start_trajectory";
 constexpr char kWriteAssetsServiceName[] = "write_assets";
 constexpr char kTrajectoryNodesListTopic[] = "trajectory_nodes_list";
+constexpr char kConstraintsListTopic[] = "constraints_list";
```
Use singular? Calling it a list already says it can contain multiple items, i.e. `kConstraintListTopic`, `constraint_list`, etc

ojura replied:
Sure. Should I fix the line above too?

wohe commented on cartographer_ros/cartographer_ros/map_builder_bridge.h:
```diff
@@ -44,32 +44,33 @@ class MapBuilderBridge {
   };
 
-  MapBuilderBridge(const NodeOptions& node_options, tf2_ros::Buffer* tf_buffer);
+  MapBuilderBridge(const NodeOptions &node_options, tf2_ros::Buffer *tf_buffer);
```
Please revert to the style used throughout the Cartographer codebase. Did you use `clang-format --style=Google`?

ojura replied:
I have, on the whole cartographer_ros codebase (and I have re-run it just now to confirm), and it outputs this. Should I revert these changes nevertheless?

jihoonl replied:
This happened to @Mofef as well. Maybe error in `clang-format`?

wohe replied:
Yes, please. Re-running `clang-format` cannot fix this issue since it changes everything in the file to match the majority style. And right now the majority of cases in this file are clearly different from the rest of the code base.

ojura replied:
Ah, I see. But it's odd that it decided to this - this style was never the majority.

wohe commented on cartographer_ros/cartographer_ros/map_builder_bridge.cc:
```diff
@@ -242,6 +244,113 @@ visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodesList() {
+
+    // Work around the 16384 point limit in RViz by splitting into
+    // multiple markers, similar to GetTrajectoryNodesList().
+    if (constraint_marker->points.size() == 16384) {
```
Did `LINE_LIST` also trigger the bug?

ojura replied:
Will test.

ojura replied:
You were right, this wasn't neccessary. Removed.

wohe commented on cartographer_ros/cartographer_ros/map_builder_bridge.cc:
```diff
@@ -242,6 +244,113 @@ visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodesList() {
+visualization_msgs::MarkerArray MapBuilderBridge::GetConstraintsList() {
+  visualization_msgs::MarkerArray constraints_list;
+  int marker_id = 0;
+  const ros::Time now = ros::Time::now();
```
This is used once: inline below?

wohe commented on cartographer_ros/cartographer_ros/map_builder_bridge.cc:
```diff
@@ -242,6 +244,113 @@ visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodesList() {
+  visualization_msgs::Marker constraint_inter_marker = constraint_intra_marker;
+  constraint_inter_marker.id = marker_id++;
+  constraint_inter_marker.ns = "Inter constraints";
+  constraint_inter_marker.pose.position.z = 0.1;
```
Why is this different for `constraint_intra_marker`?

ojura replied:
Intra constraints are numerous, while the loop closure constraints are rare and should have priority while z-fighting with intra constraints. I will add a comment.

wohe commented on cartographer_ros/cartographer_ros/map_builder_bridge.cc:
```diff
@@ -242,6 +244,113 @@ visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodesList() {
+
+  // Color mapping for submaps of various trajectories - add trajectory id
+  // to ensure different starting colors.
+  const auto submap_color_id_map =
```
Used once: inline below?

wohe commented on cartographer_ros/cartographer_ros/map_builder_bridge.cc:
```diff
@@ -242,6 +244,113 @@ visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodesList() {
+      constraint_marker = &constraint_intra_marker;
+      residual_marker = &residual_intra_marker;
+      color_constraint = GetColor(submap_color_id_map(constraint.submap_id));
+      color_residual.a = 1.0;
```
Are these colors important, i.e. why not reuse `GetColor()` here as well?

ojura replied:
How would you differentiate visually the different types of constraints and markers? Since the intra constraints are the most numerous, I decided to use the color scheme for their markers; for other types of markers, I used fixed colors (e.g. so you know that loop closure constraints are bright yellow).

wohe replied:
I'm not opposed to using fixed colors for these, I just can't see how these specific colors are better than say `GetColor(1)`, `GetColor(2)`, and `GetColor(3)`.

ojura replied:
Ah, I see. Currently, we can't generate 100% bright yellow and teal with GetColor(), as the HSV value and saturation are hard coded in colors.cc (as you made me do it :-) ) to produce a bit more muted colors. I would prefer to have loop constraints as 100% bright colors, so they are more easily spotted and differentiated from the rest.

wohe commented on cartographer_ros/cartographer_ros/map_builder_bridge.cc:
```diff
@@ -242,6 +244,113 @@ visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodesList() {
+      color_residual.b = color_residual.g = 1.0;
+    }
+
+    constraint_marker->colors.push_back(color_constraint);
```
Seems more readable to do this without the temporary variables directly above. If you do not want to repeat the `push_back` twice: loop twice around the `if`?

ojura replied:
In that case I would be assigning the pointer twice, which is very ugly to me. Is it okay if I put a for loop in each branch?

ojura replied:
Also, I cannot get rid of the temporary variables. ColorRGBA does not have a nice constructor that I could easily pass to push_back. (And I do want to retain the special bright colors for the loop closure constraints).

wohe commented on cartographer_ros/cartographer_ros/map_builder_bridge.cc:
```diff
@@ -242,6 +244,113 @@ visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodesList() {
+        all_trajectory_nodes[constraint.node_id.trajectory_id]
+                            [constraint.node_id.node_index]
+                                .pose;
+    const auto constraint_pose = submap_origin * constraint.pose.zbar_ij;
```
Let's not use `auto` in these cases. It works now, but I feel it is to easy to miss cases using Eigen types.

wohe commented on cartographer_ros/cartographer_ros/map_builder_bridge.cc:
```diff
@@ -242,6 +244,113 @@ visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodesList() {
+    const auto& submap_data =
+        all_submap_data[constraint.submap_id.trajectory_id]
+                       [constraint.submap_id.submap_index];
+    const auto& submap_origin = submap_data.pose;
```
Remove or at least rename to `submap_pose`. The origin is `submap_data.pose.translation()`.

ojura commented:
I believe it's ready for merging, I have mostly addressed the nits (and the rest is a compromise :-). Please merge #402 first.

ojura commented:
@wohe, here's some art :-)
http://free.iiifhosting.com/iiif/8cf6f3eea5aa729a5551ab535c75fce49e6a7d71fde965e4f43550d89efe8741/

ojura commented:
I have two minor questions for the reviewer:
- is it okay to use `const auto&` with e.g.  `GetAllSubmapData()` and the rest
- should I add [here](https://github.com/googlecartographer/cartographer_ros/pull/375/files#diff-aaf9b1c28824a784d19a7190cdf31e8eR320) multiplication with `node.constant_data->tracking_to_pose).translation()` like [here](https://github.com/googlecartographer/cartographer_ros/blob/master/cartographer_ros/cartographer_ros/map_builder_bridge.cc#L227)?

ojura commented:
I have just rebased on top of merged #402. Also, if you're having z-fighting-alike rendering bugs like in the video I linked above, I found that recompiling rviz with the change I described in [this comment](https://github.com/ros-visualization/rviz/issues/1067#issuecomment-312518247) helped.

wohe commented on cartographer_ros/cartographer_ros/map_builder_bridge.cc:
```diff
@@ -242,6 +244,101 @@ visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodeList() {
+  residual_inter_marker.ns = "Inter residuals";
+  residual_inter_marker.pose.position.z = 0.1;
+
+  const auto& all_trajectory_nodes =
```
`const auto` without the reference for readability since these are returned by value? Same for the submaps and constraints.

wohe commented:
Regarding the other question: Since we now only publish the translational part it does not matter. The transform I suggested adding is between yaw-only xy plane aligned frame and the full orientation. If you leave it out, the orientation will not be correct. In case you want to add it: it is needed for both the constraint pose and the trajectory node pose. You could also just add a comment why it is not needed here.

ojura commented:
In that case, would it make sense to remove it from `GetTrajectoryNodeList()` as well, because we are taking only the translational part there too? That would make things consistent. Plus, I can add a comment why it's left out.

wohe commented:
Sounds good.

ojura commented:
Done.

wohe commented:
Thanks, I'll merge as soon as the tests passed.
